proto/src/simulator/component.proto

```diff
@@ -0,0 +1,139 @@
+// This is the RPC variant of a component definition.
+//
+// It's passed over the wire during initialization, watchdogs firing,
+// or if the client explicitly requests a component summary. (E.g. for WAILA.)
+//
+// It's *not* saved to disk, as there is quite a bit of data which the client never needs. Those protobuf
+// variants exist only inside the core module.
+syntax = "proto3";
+package org.eln2.proto.simulator;
+
+// Each field is only filled in when it makes sense, i.e. a crash report will include everything,
+// as might a watchdog firing or WAILA-induced state request, but an AddComponent request won't
+// include the (unknown) electrical state. It might include thermal state.
+message Component {
+	// This value is monotonically increasing for a given server, and never reused.
+	int64 id = 1;
+
+	// This is the computed electrical and thermal state, usually sent server->client.
+	ElectricalState electrical_state = 2;
+	ThermalState thermal_state = 3;
+
+	// These are initialization parameters, sent client->server.
+	Watchdog watchdog = 4;
+	ThermalParams thermal = 5;
+	Location location = 6;
+	// In general, default to adding components to this list (handling them core-side) rather than
+	// emulating them client-side. E.g, a lightbulb is better simulated in core than on the client.
+	// This is both for performance (it's much easier to multithread Core than a minecraft server),
+	// and for code de-duplication.
+	oneof type {
+		// Circuit primitives:
+		VoltageSource voltage_source = 10;
+		CurrentSource current_source = 11;
+		Resistor resistor = 12;
+		Inductor inductor = 13;
+		Capacitor capacitor = 14;
+		// Cables:
+		Resistor simple_cable = 20;
```

Comment:
Cables are really common and should probably be under ID 16 since it compresses the data in serialization.

Reply:
Also, is there really a reason to differentiate the cable from a standard resistor, since we would likely use line for both?

Reply:
TBH, that was supposed to be an example of the 'default to defining new components' rule. I later found a better one. For now I'll just remove it, but if cables end up doing *anything* different than resistors they should be a separate component type.

Reply:
I can't currently see why cables would be any different from resistors as far as the MNA is concerned. In the thermal sim, yes, there will be a difference or two, but our cables are not going to become thermistors (resistors that change resistance as a result of temperature) because that would be prohibitively complicated to process and cause many unnecessary matrix recalcs on a per-thermal-tick level. Also, cables will be of a floating point length due to poles and other non-block cable types. Plus, I intend to introduce materials with different resistances and such.. but I don't think any of that matters to the MNA and as such doesn't need to be sent - it only needs to know about the resistance I think.